server/study_server.js

```diff
@@ -153,56 +153,59 @@ module.exports = {
-                            req.end();
-                        } else {
-                            if (response.complete){
-                                res.redirect('/msg/nomore');
-                                res.end();
-                            } else {
-                                 Response.findOneAndUpdate({"_id": clean_resid},
-                                    { "$set": { "complete": true,
-                                                "date": new Date(Date.now()),
-                                                "data": JSON.parse(req.body.result)}
-                                    }).exec(function(err, book){
-                                       if(err) {
-                                           console.log(err);
-                                           res.status(500).send(err);
-                                       }
-                                });
-                                //move response object from incompleteResponses to completeResponses
-                                var respIdx = study.incompleteResponses.indexOf(req.body.resid);
-                                study.completeResponses.push(study.incompleteResponses[respIdx]);
-                                study.incompleteResponses.splice(respIdx,1);
-                                //save the study object (which will save the child objects)
-                                study.save();
-                                res.redirect('/msg/thanks');
-                                res.end();
-                            }
-                        }
-                    });
+        function updateStudy() {
+            Study.findOneAndUpdate(
+                { "_id": clean_id },
+                {
+                    "$pull": { "incompleteResponses": mongoose.Types.ObjectId(clean_resid) },
+                    "$push": { "completeResponses": clean_resid }
+                },
+                function (err) {
+                    if (err) {
+                        logger.error("study_server.js: Error has occured: ", err);
+                        res.status(500).send(err);
+                        res.end();
+                    } else {
+                        logger.info("study_server.js: Response saved successfully to study.");
+                        res.redirect('/msg/thanks');
+                        res.end();
+                    } 
+                }
+            );
+    }
 
+        function saveResponse(){
+            Response.findOne({_id: clean_resid}).then(response => {
+                if (!response) {
+                    logger.error("study_server.js: Response not found");
+                    res.status(404).send("Response not found");
+                    res.end();
+                }
+                if(response.complete){
+                    logger.error("study_server.js: The response has already been completed");
```

Comment:
Previous code redirected to msg page if the response is complete. Let's not change that